src/map/battle.cpp

```diff
@@ -10794,44 +10794,138 @@ enum damage_lv battle_weapon_attack(struct block_list* src, struct block_list* t
+			if(skill_lv == 0)
+				return;
+			int32 i = rnd()%100;
+			if(sc->getSCE(SC_SPIRIT) && sc->getSCE(SC_SPIRIT)->val2 == SL_SAGE)
+				i = 0; //Max chance, no skill_lv reduction. [Skotlex]
 		//reduction only for skill_lv > 1
-		if (skill_lv > 1) {
-			if (i >= 50) skill_lv /= 2;
-			else if (i >= 15) skill_lv--;
-		}
-		sp = skill_get_sp(skill_id,skill_lv) * 2 / 3;
+			if(skill_lv > 1){
+				if(i >= 50) skill_lv /= 2;
+				else if(i >= 15) skill_lv--;
+			}
+#else
+			uint16 autospell_level{};
+			if(sd){
+				autospell_level = pc_checkskill(sd,SA_AUTOSPELL);
 
-		if (status_charge(src, 0, sp)) {
-			struct unit_data *ud = unit_bl2ud(src);
+				if(autospell_level == 0)
+					return;
 
-			switch (skill_get_casttype(skill_id)) {
-				case CAST_GROUND:
-					skill_castend_pos2(src, target->x, target->y, skill_id, skill_lv, tick, flag);
-					break;
-				case CAST_NODAMAGE:
-					skill_castend_nodamage_id(src, target, skill_id, skill_lv, tick, flag);
-					break;
-				case CAST_DAMAGE:
-					skill_castend_damage_id(src, target, skill_id, skill_lv, tick, flag);
-					break;
+				if(rnd()%100 >= (5 + autospell_level * 2))
+					return;
+			}else{
+				if(rnd()%100 >= sc->getSCE(SC_AUTOSPELL)->val4)
+					return;
 			}
-			if (ud) {
-				int32 autospell_tick = skill_delayfix(src, skill_id, skill_lv);
+			uint16 skill_lv{};
+			int32 sp{};
+
+			uint16 skill_id = sc->getSCE(SC_AUTOSPELL)->val2;
+			if(skill_id == 0)
+				return;
 
-				if (DIFF_TICK(ud->canact_tick, tick + autospell_tick) < 0) {
-					ud->canact_tick = i64max(tick + autospell_tick, ud->canact_tick);
-					if (battle_config.display_status_timers && sd)
-						clif_status_change(src, EFST_POSTDELAY, 1, autospell_tick, 0, 0, 0);
+			uint16 max_level = sd ? pc_checkskill(sd,skill_id) : sc->getSCE(SC_AUTOSPELL)->val3;
+
+			if(max_level == 0)
+				return;
+
+			if(!sd || (sd && sd->sc.getSCE(SC_SPIRIT) && sd->sc.getSCE(SC_SPIRIT)->val2  ==  SL_SAGE))
+				skill_lv = max_level;
+			else{
+				switch(skill_id){
+					case MG_NAPALMBEAT:
+						skill_lv = 3;
+						break;
+					case MG_COLDBOLT:
+					case MG_FIREBOLT:
+						[[fallthrough]];
+					case MG_LIGHTNINGBOLT:
+						if(autospell_level == 2)
+							skill_lv = 1;
+						else if(autospell_level == 3)
+							skill_lv = 2;
+						else if(autospell_level >= 4)
+							skill_lv = 3;
+						break;
+					case MG_SOULSTRIKE:
+						if(autospell_level == 5)
+							skill_lv = 1;
+						else if(autospell_level == 6)
+							skill_lv = 2;
+						else if(autospell_level >= 7)
+							skill_lv = 3;
+						break;
+					case MG_FIREBALL:
+						if(autospell_level == 8)
+							skill_lv = 1;
+						else if(autospell_level >= 9)
+							skill_lv = 2;
+						break;
+					case MG_FROSTDIVER:
+						skill_lv = 1;
+						break;
+					default:
+						return;
+				}
+
+				skill_lv = min(max_level,skill_lv);
+				if(skill_lv > 1){
+					uint32 i = rnd()%1000;
+					switch(skill_lv){
+						case 2:
+							if(i >= 235)
```

Comment:
How did you conclude these numbers? Seems overcomplicating it compared to how it works on Aegis. At least explain in a comment how the math works here.